### PR TITLE
[v6r15] Performance update in DataStoreClient

### PR DESCRIPTION
**Pull request by hgiemza**
* Synchronizer based decorators have been replaced with a simple lock as they were blocking addRegister() during every commit() (and commits could take several seconds to finish)
* Added delayedCommit() which reduces overheads. Call to this function will in the future be propagated to BaseAccountingType class and similar places
* See comment in the lines 59-60
* Change of the target branch (see pull request #2747)

### CODE REVIEW

**Comment by fstagni** [comment]
Review is OK, but I would avoid putting delayedCommit() in this PR if it's not used for now.

File: AccountingSystem/Client/DataStoreClient.py [hunk 1]
```diff
@@ -58,7 +55,13 @@ def addRegister( self, register ):
       return retVal
     if gConfig.getValue( '/LocalSite/DisableAccounting', False ):
       return S_OK()
+    
+    # a lock below can only prevent to send the register 
+    # with an 'earlier' commit what normally is not a problem at all
```
Comment (chaen) [inline AccountingSystem/Client/DataStoreClient.py]: For my understanding, can you just explain what you mean with "normally is not a problem at all"? Do you mean that we don't care if they are in different order, or do you mean that it should not happen even without the lock ?

File: AccountingSystem/Client/DataStoreClient.py [hunk 2]
```diff
@@ -69,29 +72,53 @@ def __getRPCClient( self ):
...
+    
+    # create a local reference and prevent other running commits
+    # to take the same data second time 
+    self.__registersListLock.acquire()
+    registersList = self.__registersList
+    self.__registersList = []
+    self.__registersListLock.release()
+    
+    try:
+      while len( registersList ) > 0:
```
Comment (chaen) [inline AccountingSystem/Client/DataStoreClient.py]: while registersList ?

File: AccountingSystem/Client/DataStoreClient.py [hunk 3]
```diff
@@ -69,29 +72,53 @@ def __getRPCClient( self ):
...
+            result = _sendToFailover( retVal[ 'rpcStub' ] )
+            if not result[ 'OK' ]:
+              return result
+          else:
+            return S_ERROR( "Cannot commit data to DataStore service" )
+        sent += len( registersToSend )
+        del( registersList[ :self.__maxRecordsInABundle ] )
+    finally:
+      # if something is left because of an error return it to the main list
+      self.__registersList.extend(registersList)
```
Comment (chaen) [inline AccountingSystem/Client/DataStoreClient.py]: Don't you need the lock here ?

Comment (hgiemza) [inline AccountingSystem/Client/DataStoreClient.py]: If you mean self.__registersList.extend(registersList) then, as far as I know, it doesn't need an additional lock. Please see here for an explanation:
http://effbot.org/pyfaq/what-kinds-of-global-value-mutation-are-thread-safe.htm

File: AccountingSystem/Client/DataStoreClient.py (hunk 2 above)
Comment (hgiemza) [inline AccountingSystem/Client/DataStoreClient.py]: Good point I will change.

File: AccountingSystem/Client/DataStoreClient.py (hunk 1 above)
Comment (hgiemza) [inline AccountingSystem/Client/DataStoreClient.py]: Please consider this part in context of the previous code and the way it worked.

The old version of commit() 'permits' occasional earlier commits of registers added by another threads as it uses the same internal variable (self.__registersList) to get the list of elements which should be send to the database. Having in mind this feature I've assumed that "an 'earlier' commit is not a problem at all" as the previous code worked this way.

Considering the order of data we also have similar behaviour in both cases. Previously, if we had 2 or more threads that were executing addRegister() simultaneously using the same client object (e.g. gDataStoreClient), we didn't have any influence on the final order of elements in the internal register (self.__registersList) and thus in database.

**Comment by hgiemza** [comment]
I've introduced both changes, so it should be ready to merge now.

**Comment by chaen** [comment]
Review OK

**Pull request merged**